### DISCUSSION TITLE
Request for development workflow for library on novice with pyscaffold

### DISCUSSION THREAD

**Discussion by Seersucker** [pyscaffold/pyscaffold#368] [Q&A] [answered]
Hello,
Thanks for all the hard work creating pyScaffold! I've started and gave up multiple times when attempting to develop a python library. I'm able to create some source code, execute unit tests and save changes to my repo. After a while, I start getting problems since I it might be weeks or more since this is a hobby. What is the recommended workflow for going from proof-of-concept python script to module or application. I see the option for "python setup.py develop" but not sure where I'm lost.  When I reboot my laptop, and run pip list or freeze, I don't see my module I've created with pyScaffold.  Please consider expanding the documentation with a tutorial on how to create a pyScaffold project from start to finish.  My general workflow is:

1. pip install pyscaffold
2. cd repos
3. putup mytest1 --pre-commit -d 'My proof of concept' --tox --travis
3. cd mytest1
4. vim setup.cfg -> add packages I know I'll use
5. pipenv install -e .  (setup venv inside my pyscaffold project)
6. pipenv install --dev tox travis sping pytest
7. create some simple python script in src/mytest1
8. commit to git and tag
9. python setup.py develop
10. get stuck here

USE CASE: I want to try out some of the module that I've created in ipython as a sanity check. When I look at pip list, I do see my "pytest1 listed with a git tag (e.g. v0.0.2).

Do most people develop their proof of concept code outside of pyscaffold; and, only after they have a good working example, work to port it into pyScaffold? Any help is appreciated.

I found this video that shows the workflow for a pyScaffold command line tool executable but I'm interested how to develop properly for a package library. 

https://www.youtube.com/watch?v=JwwlRkLKj7o

**Comment by FlorianWilhelm** [comment, marked as answer]
Hi @Seersucker, 

thanks for your feedback and for enumerating the steps in your workflow, this helps us a lot in figuring out what the problem is. In general, my advice would be to start with creating a project setup with PyScaffold before you even start coding on a new project. At least that's how I proceed even when I am doing a PoC. Since I mostly work on Data Science projects my workflow is:

1. `pip install pyscaffold pyscaffoldext-dsproject`   # those are actually installed in my base conda environment 
2. `cd Sources`  # my directory for holding all my projects and repos
3. `putup --dsproject my_ds_project`
4. `cd my_ds_project`
5. `conda env create -f environment.yml` # to create an own conda environment for my new project
6. `conda activate my_ds_project`  # activates the environment just created
7. `python setup.py develop`  #  or  alternatively `pip install -e .` to install `my_ds_project` in the environment
8. `jupyter lab` # to work on jupyter notebooks
9. Hack away on some Jupyter notebook until something works how I want it or the size of the notebook becomes too large
10. If I need an external library (production dependency) I add it to `setup.cfg` and also to `environment.yml`(here also development dependencies like `py.test`, `pre-commit` and others).
11. Extract working parts from the notebook by creating a module, e.g. `src/my_ds_project/model.py`, and move the code over using PyCharm. From this point on I test things interactively in the notebook, but also do a huge part of my codings in PyCharm directly.
12. Create a unit test, e.g. `tests/test_model.py` and run it with `py.test` or `tox` 
13. Call `python setup.py develop` or `pip install -e .` again since we created a new module. That's not necessary if an existing module was just extended.
14. Import the model module in my notebook with `import model` and start using it in the notebook
15. Commit what I have.
16. **Loop:** Repeat steps 10 to 15 over quite some time. It's important to remember after a long break to activate the conda environment again. 
17. When reaching a first milestone in my project, I tag it and start thinking about setting up a CI/CD pipeline (e.g. Cirrus, Gitlab/Github) and where to publish my packages, maybe PyPI or Artifcatory (if it is a customer project). If I am working in a team on a project directly from the start, this step has to come much earlier. 

Judging from your experience,  you might forget to activate your pipenv after step 7. This is also necessary again after you took a break from your project, i.e. every time you closed the shell. I actually never used `pipenv` since I think it does too much magic and rather recommend Conda but many people also love pipenv.

Does this help you?

**Reply by Seersucker** [reply]
Hello Wilhelm,
Yes, this is very helpful! Danke! So, after long breaks on my toy project, I would either forget to activate the virtual environment or load the new module into memory.  

So, if I don't see the virutal environment name in my terminal and the project module in my pip list, I know what to do:

1. pipenv shell ( or conda activate mytest1)
2. python setup.py develop

## Suggested Workflow Ideas

I started a script to accomplish steps 1 -> 7. Does it make sense to share or integrate into pyScaffold? Consider having the `putup` cmd returning a virtual environment with the project loaded into its virtual python packages. This would allow novices to start using it without polluting their OS.  The `putup` cmd could print out instructions on how to activate everything before it returned ownership back to the console.  

## Suggested Documentation Edits

On the first page of [documentation](https://pyscaffold.org/en/latest/) 

I see `pip install -e .` it is not obvious what is occurring. 

It would be helpful to describe any command with a brief description "this installs your my_project module in the current Python environment. FYI, be sure to check which Python environment you are in before executing: either local or a virtual one."

It might not be obvious to a Python novice that using a virtual environment is good practice. I would recommend including a "how-to" to prevent them from polluting their default Python installation. 

### FAQ or Sections for Review

- How to Choose a Virtual Environment for your Project
- How to Create a Virtual Environment
- How to Activate your Virtual Environment
  - How to Develop Your Project within your Virtual Environment
- How to Deactivate your Virtual Environment



#### Options for Activating your Virtual Environment

Show how a terminal (e.g., zsh) would change to indicate the active, virtual environment

#### Using venv
```
user1@localhost:
user1@localhost: source my_project/bin/activate
(venv) user1@localhost:
```

#### Using conda

```
user1@localhost
user1@localhost: conda activate my_project
(my_project) user1@localhost:
```

#### Using pipenv

```
user1@localhost
user1@localhost: pipenv shell
(my_project) user1@localhost:
```

## OS Choice Impacts Development Reflection

Ideally, I would have a Docker file that would create a Python development that was portable and lean. So, no matter what box I'm on, its the same.

Through this process, I learned Python development on Windows10 , including WSL2, is a major hassle and should be avoided.  Developing on my old laptop with Ubuntu always worked!  Unfortunately, pyCharm Community edition does not allow attachment to docker for remote runtimes. While VSCode supports Python development in a remote runtime in Docker container, there are far too many setup steps required.

**Reply by abravalheri** [reply]
Hi @Seersucker I did some small improvements on the latest docs based on your input (while I wait for the time to properly address the other suggestions), I hope it is a bit more clear now. There are 2 things that you might also find interesting:

1. PyScaffold v4 (currently in development) has a [`--venv`](https://pyscaffold.org/en/latest/dependencies.html#development-environment) flag that will automatically create a virtualenv and display some quick help on how to activate/use it once `putup` finishes run. I suppose these are the automation steps you have mentioned in your last comment? If you are a 1 person developer team, that should be enough to get you far, otherwise you might be interested the following:
2. We have [some docs on how to integrate dependency managers](https://pyscaffold.org/en/latest/dependencies.html#how-to-integrate-pipenv), maybe that helps? There the advice is to use `pipenv run` to run commands and bypass the complications of having to always activate the virtual environment.

Personally, I got used to use the `--venv` option and always type `.venv/bin/{python,pip,...}`, it is a bit more verbose, but I am always sure to be using the correct virtual environment (moreover `CTRL+R` on the terminal usually saves most of the typing anyway ...).

One final comment about Windows 10: I use WSL + virtualenv at work and I never had any problems, PyScaffold works just fine... Is there any particular problem you face? Maybe I can help.

**Reply by Seersucker** [reply]
Hello Anderson,
Thanks for the reply! I will try out the new v4 version of PyScaffold. 

 

I did notice that when I installed Anaconda in WSL it was possible to get confused whether the virutal conda or venv (or pipenv) was activated. I typically use the ZSH. I discovered that conda environments don't live in PyScaffold. Do you typically have separate environments (not sure how to accomplish this other than Docker) when bouncing between conda and others? I guess it makes sense to stick to one and leave everything else alone. It just seems the bloat of a new conda is overkill for experimentation.
 
  

It is probably time to reinstall Windows 10 since many strange issues.  Cheers from across the pond.

**Reply by abravalheri** [reply]
Hi @Seersucker, thanks for the feedback. I hope you manage to find the right workflow for you.

I tend to always use a virtual environment per project, it is makes life easier. But I guess you could get away if you have just one project or if your projects don't have dependencies...

Conda does take a while to create a new environment, that is why on a daily basis I use [virtualenv](https://virtualenv.pypa.io/en/latest/), which in my opinion is the fastest (I believe it creates symlinks to the binaries on the system so not too overkill). To be sincere I use conda only in 3 different situations:

a) number crunching/data science projects
b) when I have loads of dependencies that encapsulate C/C++ libraries (like GUIs or industrial equipment automation)
c) to install a recent version of Python in operating systems that don't have official packages for it yet (like Python 3.9 on Ubuntu 18.04)

In the case of (a) and (b), I think the wait is totally worthy... It is better to use conda then face dependency hell later. In the case of (c), I suppose one could try a combination of conda and virtualenv (install the desired python version with conda and then install/use virtualenv from inside the python distribution given by conda -- but I never tried).

There are a few articles that can help if you are new to using isolated environments in Python:
- https://realpython.com/python-virtual-environments-a-primer/
- https://towardsdatascience.com/virtual-environments-104c62d48c54
- https://towardsdatascience.com/a-guide-to-conda-environments-bc6180fc533

But in the end of the day what matters is to be happy and productive with your workflow, so if you are already familiar with docker, just go for it 😝
